core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MakeAsyncHttpRequestStage.java

```diff
@@ -188,17 +214,44 @@ public void exceptionOccurred(Throwable throwable) {
      */
     private class Completable {
         private final CompletableFuture<Response<OutputT>> completableFuture = new CompletableFuture<>();
+        private AsyncTimeoutTask timeout;
+        private TimeoutTracker timeoutTracker;
+
+        Completable(long timeoutInMills) {
+            if (isApiCallAttemptTimeoutEnabled(timeoutInMills)) {
+                timeout = new AsyncTimeoutTask(completableFuture, ApiCallAttemptTimeoutException.create());
+
+                ScheduledFuture<?> scheduledFuture = timeoutExecutor.schedule(timeout, timeoutInMills, TimeUnit.MILLISECONDS);
+                timeoutTracker = new ApiCallTimeoutTracker(timeout, scheduledFuture);
+
+                completableFuture.whenComplete((o, t) -> {
+                    if (t != null) {
+                        timeoutTracker.cancel();
+                    }
+                });
+            } else {
+                timeoutTracker = new NoOpTimeoutTracker();
+            }
+        }
 
-        public void complete(Response<OutputT> result) {
+        void abortable(Abortable abortable) {
+            if (timeout != null) {
+                timeout.abortable(abortable);
+            }
+        }
+
+        void complete(Response<OutputT> result) {
             try {
+                log.trace(() -> "completing the result");
                 futureCompletionExecutor.execute(() -> completableFuture.complete(result));
             } catch (RejectedExecutionException e) {
                 completableFuture.completeExceptionally(explainRejection(e));
             }
         }
 
-        public void completeExceptionally(Throwable exception) {
+        void completeExceptionally(Throwable exception) {
             try {
+                log.trace(() -> "completeExceptionally");
```

Comment:
Still needed? Same above.

Reply:
nope, I have a `clean up logs` TODO items in the description. :) will remove them when it's close to merge.